### PR TITLE
feat: add @mastra/aisearch - Azure AI Search vector store

### PR DESCRIPTION
**Pull request by valdepeace**
## Review Summary

This is an excellent addition to the Mastra project - a complete implementation of an adapter for Azure AI Search as a vector store. I reviewed the code and have some general comments:

### Positive Aspects ✅

1. **Complete Implementation**: The PR includes a fully functional implementation of the Azure AI Search Vector Store client with:
2.    - Basic vector operations (upsert, query, delete)
3.    - Support for advanced search (semantic, hybrid, HNSW tuning)
4.    - Seamless integration with Mastra Memory
2. **Comprehensive Documentation**: The README is very complete with:
3.    - Clear installation guide
4.    - Practical usage examples
5.    - Details about prerequisites
6.    - Step-by-step configuration
3. **Test Coverage**: Includes extensive tests:
4.    - Unit tests with mocked SDK
5.    - Optional integration tests
6.    - Edge case coverage
4. **Flexible Configuration**: The client supports:
5.    - API Key and Azure authentication
6.    - Customizable ClientOptions
7.    - Custom proxies and headers
### Observations 📝

1. **Merge Status**: The PR is marked as blocked - it needs at least 1 approval from a reviewer with write access before it can be merged.
2. **Pending Tests**: There are 3 checks still awaiting:
3.    - Combined store Tests
4.    - E2E Tests
5.    - Full Test Suite
3. **Outdated Branch**: The branch is out-of-date with the base branch. It's recommended to update to avoid merge conflicts.
4. **OData Field Validation**: CodeRabbit flagged an important issue in filter.ts regarding `escapeFieldName` - Azure AI Search requires field paths without quotes for certain query types.
### Recommendations

- [ ] Wait for all test checks to complete
- [ ] - [ ] Update the branch with the latest changes
- [ ] - [ ] Review CodeRabbit's comment about OData syntax
- [ ] - [ ] Ensure at least one project maintainer approves the changes
Excellent work on the implementation! This vector store will open many possibilities for Azure AI Search users in Mastra.

### CODE REVIEW

**Comment by abhiaiyer91** [comment]
Thank you for this PR @valdepeace ! Please address the coderabbit review!

**Comment by abhiaiyer91** [comment]
This is looking great @valdepeace !!

Can we change the name of this to be more explicit like `stores/azure`

**Comment by valdepeace** [comment]
Hi @abhiaiyer91! Thanks for the feedback 🙌

I see what you mean about being more explicit.

Right now, each store in the repo follows a pretty simple pattern:
- one folder under `stores/` → one package (e.g. `stores/astra` → `@mastra/astra`, `stores/chroma` → `@mastra/chroma`).

If we rename the folder to `stores/azure` but keep the package as `@mastra/aisearch`, it works, but the folder/package naming becomes a bit inconsistent compared to the other stores.

To keep it explicit **and** consistent with the existing pattern, I’d suggest:

- folder: `stores/azure-aisearch`
- package: `@mastra/azure-aisearch`

That makes it clear this is specifically for **Azure AI Search**, and still follows the “one folder = one store = one package” convention.

If you’d still prefer `stores/azure` specifically, I can also:
- rename the folder to `stores/azure`
- keep the package name as `@mastra/aisearch` (or update it to whatever naming you think fits best)

Happy to align with whatever convention you want for provider-specific stores — just let me know your preference and I’ll update the PR. 🙌

File: stores/azure/src/vector/index.ts [hunk 1]
```diff
@@ -0,0 +1,1217 @@
...
+    indexName,
+    vectors,
+    metadata = [],
+    ids,
+    deleteFilter,
+  }: AzureAISearchUpsertParams): Promise<string[]> {
+    try {
+      if (deleteFilter) {
+        await this.deleteVectors({ indexName, filter: deleteFilter });
+      }
```
[the remark this answers is not in the document]
Comment (wardpeet) [reply stores/azure/src/vector/index.ts:710]: Claude says: Thanks for flagging the atomicity concern. The mergeDocuments/deleteDocuments result checking has been addressed in commit 9bd0402.

Regarding the deleteFilter atomicity: Azure AI Search does not support transactional operations across document batches. There is no way to make a delete+insert atomic on this platform — batch operations are inherently non-transactional. This is consistent with how other Mastra vector store adapters (e.g., Pinecone, Qdrant, Chroma) handle deleteFilter in upsert: they all perform sequential delete-then-insert because the underlying platforms do not provide cross-batch transactions.

The core type comment is aspirational rather than a strict contract enforced by all adapters. Adding a compensating rollback (re-insert on upload failure) would introduce significant complexity and its own failure modes (what if the rollback itself fails?). I think the current behavior is the pragmatic choice for this platform.

— Claude

**Comment by rakeshgajula** [comment]
It would be great if this PR can be merged

**Comment by valdepeace** [comment]
## Review Summary

This is an excellent addition to the Mastra project: the implementation of a complete adapter for Azure AI Search as a vector store. I have reviewed the code and have some general comments:

## Positive Aspects ✅

**Complete Implementation:**
The PR includes a fully functional implementation of the Azure AI Search Vector Store client with:

* Basic vector operations: upsert, query, delete
* Support for advanced search: semantic, hybrid, HNSW tuning
* Smooth integration with Mastra Memory

**Comprehensive Documentation:**
The README is very complete and includes:

* A clear installation guide
* Practical usage examples
* Details about prerequisites
* Step-by-step configuration

**Test Coverage:**
Extensive tests are included:

* Unit tests with a mocked SDK
* Optional integration tests
* Edge case coverage

**Flexible Configuration:**
The client supports:

* API Key and Azure authentication
* Customizable ClientOptions
* Proxies and custom headers

## Observations 📝

**Merge Status:**
The PR is marked as blocked. It needs at least one approval from a reviewer with write permissions before it can be merged.

**Pending Tests:**
There are 3 checks still waiting:

* Combined store Tests
* E2E Tests
* Full Test Suite

**Outdated Branch:**
The branch is out of date with the base branch. It is recommended to update it to avoid merge conflicts.

**OData Field Validation:**
CodeRabbit pointed out an important issue in `filter.ts` regarding `escapeFieldName` that should be reviewed. Azure AI Search requires field paths without quotes for certain types of queries.

## Recommendations

* Wait for the test checks to finish
* Update the branch with the latest changes
* Review CodeRabbit’s comment about OData syntax
* Make sure at least one project maintainer approves the changes

Excellent work on the implementation! This vector store will open up many possibilities for Azure AI Search users in Mastra.

**Comment by valdepeace** [comment]
> It would be great if this PR can be merged

Thanks a lot for the support! I agree, this would be really useful for Mastra users working with Azure AI Search.

I’ve addressed the main review feedback and added the latest fixes, so now I’m waiting for the remaining checks and maintainer review. Any additional feedback is very welcome.

**Comment by valdepeace** [comment]
Quick update on the latest push.

  While rebasing/updating this PR I noticed that `@mastra/azure` now already exists in `workspaces/azure` for the Azure Blob Storage workspace filesystem and
  blob store support.

  Because this PR was also introducing a separate package with the same npm name, keeping it as `@mastra/azure` would create two workspace packages with the
  same package name:

  - `workspaces/azure` -> `@mastra/azure`
  - `stores/azure` -> previously also `@mastra/azure`

  That makes package selection ambiguous in pnpm and would conflict during release/publishing.

  To avoid that collision while keeping Azure AI Search in the vector-store area of the repo, I renamed the package to:

  ```ts
  import { AzureAISearchVector } from '@mastra/azure-ai-search';

  The implementation remains under stores/azure, since Azure AI Search is being added here as a vector store provider, not as a workspace filesystem provider.

  I also updated the README, changelog, and changesets to use @mastra/azure-ai-search. The changeset bot now detects the new package name correctly.

  Local validation passed:

  pnpm --filter @mastra/azure-ai-search test:unit
  pnpm --filter @mastra/azure-ai-search build
  pnpm --filter @mastra/azure-ai-search lint
```
  I think this keeps the package explicit, avoids colliding with the existing Azure Blob package, and preserves the repo structure for vector stores.

**Comment by valdepeace** [comment]
Hi @abhiaiyer91, sorry for the ping.

I saw you’re assigned to this PR, so I wanted to ask for guidance on the next step:

https://github.com/mastra-ai/mastra/pull/10146

The PR adds Azure AI Search as a native Mastra vector store provider, currently as `@mastra/azure-ai-search`.

I’m happy to keep working on it, but before doing more cleanup I’d like to confirm whether this is aligned with Mastra’s roadmap.

Main question:

1. Should this stay as a separate package: `@mastra/azure-ai-search`
2. Should it move under the existing Azure package structure, for example `@mastra/azure/ai-search`
3. Or should I close the PR if native Azure AI Search support is not something Mastra wants to support right now?

I also created a small demo repo to validate the use case:

https://github.com/valdepeace/mastra-azure-aisearch-demo

If the team is interested, I’m happy to rebase on latest `main`, address the remaining CodeRabbit comments, reduce the initial scope if needed, and help maintain the package over time.

Thanks!

**Comment by valdepeace** [comment]
Hi @abhiaiyer91, quick status update:

  I’ve addressed the outstanding CodeRabbit feedback and re-triggered review. The package is currently named `@mastra/azure-ai-search` to avoid colliding with the existing `@mastra/azure` package in
  `workspaces/azure`.

  Local validation:
  - `pnpm --filter @mastra/azure-ai-search test:unit`
  - `pnpm --filter @mastra/azure-ai-search build`
  - `pnpm --filter @mastra/azure-ai-search lint`

  The only decision I still need from the team is whether Azure AI Search should ship as:
  1. `@mastra/azure-ai-search`
  2. part of an existing Azure package structure
  3. or not be accepted as a native provider right now

  Once that direction is confirmed, I can make the final naming/structure adjustment and keep the package maintained.

**Comment by wardpeet** [comment]
pr is a bit too old, let's close this out. feel free to reopen.
